crates/db_schema/src/traits.rs

```diff
@@ -1,43 +1,144 @@
 use crate::{
   newtypes::{CommunityId, DbUrl, PersonId},
-  utils::DbPool,
+  utils::{get_conn, DbPool},
 };
-use diesel::result::Error;
+use diesel::{
+  associations::HasTable,
+  dsl,
+  expression::{AsExpression, TypedExpressionType},
+  expression_methods::ExpressionMethods,
+  insert_into,
+  query_builder::{DeleteStatement, InsertStatement, IntoUpdateTarget},
+  query_dsl::methods::{FindDsl, LimitDsl},
+  result::Error,
+  sql_types::SqlType,
+  AsChangeset,
+  Expression,
+  Insertable,
+  Table,
+};
+use diesel_async::{
+  methods::{ExecuteDsl, LoadQuery},
+  AsyncPgConnection,
+  RunQueryDsl,
+};
+use futures_util::{Future, TryFutureExt};
+use std::{hash::Hash, pin::Pin};
+
+/// Returned by `diesel::delete`
+pub type Delete<T> = DeleteStatement<<T as HasTable>::Table, <T as IntoUpdateTarget>::WhereClause>;
+
+pub type Find<'a, T> = dsl::Find<<T as HasTable>::Table, <T as Crud<'a>>::IdType>;
 
+pub type InsertValues<'a, 'b, T> =
+  <&'a <T as Crud<'b>>::InsertForm as Insertable<<T as HasTable>::Table>>::Values;
+
+// When using `RunQueryDsl::execute`, directly building futures with `Box::pin` and `TryFutureExt::and_then`
+// instead of `async` + `await` fixes weird compile errors.
+// https://github.com/rust-lang/rust/issues/102211
+// When using `RunQueryDsl::first` or 'RunQueryDsl::get_result`, `async` + `await` works, and it must be used otherwise the closure for `and_then`
+// will both own `conn` and return a future that references it.
 #[async_trait]
-pub trait Crud {
-  type InsertForm;
-  type UpdateForm;
-  type IdType;
-  async fn create(pool: &mut DbPool<'_>, form: &Self::InsertForm) -> Result<Self, Error>
+pub trait Crud<'a>
```

Comment:
Just because this compiles, doesn't mean we should do it. This is an unmaintainable mess of typings IMO

Reply:
I'm going to try moving the lifetime parameter to the associated types. Hopefully it will remove the need for messy workarounds for the lifetime issue.

And maybe the trait bounds can be better organized by moving them to the associated types and/or using comments.